vignettes/slowly-changing-dimension.Rmd

```diff
@@ -0,0 +1,241 @@
+  "Seattle",     16, "2023-09-28",
+  "Houston",     34, "2023-09-28",
+  "New York",    18, "2023-09-29",
+  "Los Angeles", 25, "2023-09-29",
+  "Seattle",     17, "2023-09-29",
+  "Houston",     34, "2023-09-29"
+) %>%
+  mutate(across("ForecastDate", as.Date))
+
+forecasts_scd <- forecasts_full %>%
+  mutate(ForecastFrom = ForecastDate,
+         ForecastUntil = case_when(
+           City == "Houston" ~ NA,
+           ForecastDate == "2023-09-29" ~ NA,
+           TRUE ~ "2023-09-29"
+         )) %>%
+  mutate(across("ForecastUntil", as.Date)) %>%
+  filter(!(City == "Houston" & ForecastFrom == "2023-09-29")) %>%
+  distinct() %>%
+  select(!"ForecastDate") %>%
+  arrange(ForecastFrom)
+
+# Only define one connection
+if (!exists("con")) con <- DBI::dbConnect(drv = RSQLite::SQLite(), dbname = ":memory:")
+
+forecasts_dated <- filter(forecasts_full, ForecastDate == "2023-09-28")
+forecasts <- select(forecasts_dated, !"ForecastDate")
+
+forecasts2 <- filter(forecasts_full, ForecastDate == "2023-09-29") %>%
+  select(!"ForecastDate")
+forecasts2a <- filter(forecasts_full, ForecastDate == "2023-09-29")
+```
+
+A slowly changing dimension is a concept in data warehousing which refers to data which may change over time, but at an irregular schedule.
+
+## Type 1 and Type 2 history
+
+For example, consider the following table of forecasts for a number of cities:
+```{r forecasts}
+# Current date: 2023-09-28
+forecasts
+```
+
+The following day, the forecasts will have changed, and — barring the occasional data hoarder — the existing data is no longer relevant.
+
+In this example, most (if not all) of the values of the `Forecast_C` column will change with each regular update.
+Putting it into other words, the table is a *snapshot*^[A *snapshot* is a static view of (part of) a database at a specific point in time] of forecasts at the last time of update.
+
+The following day, the forecasts naturally change:
+```{r forecasts2a}
+# Current date: 2023-09-29
+forecasts2
+```
+We could choose to update the forecasts table so that it would always contain the current data.
+This is what is referred to as Type 1 methodology [@Kimball2013].
+
+Databases are thankfully a rather efficient way of storing and accessing data, so instead of discarding the values from the previous day, we append the new data to those of the previous day.
+Also, in order to keep our data organised, we add a column with the date of the forecast, aptly named `ForecastDate`.
+
+The full table of forecasts for the two days now looks like below, and we are slowly building a full history of forecasts:
+```{r forecasts_full}
+forecasts_full
+```
+Managing historical data by inserting new data in this manner is often referred to as Type 2 methodology or Type 2 history.
+
+Our table now provides much more information for the user through filtering:
+```{r forecasts_full_examples}
+library(tidyverse)
+
+# Current forecasts
+forecasts_full %>%
+  slice_max(ForecastDate, n = 1) %>%
```

Comment:
`n = 1` is used by default, so you don't _need_ to specify it if you don't want to